### DISCUSSION TITLE
QUESTION: how to do theory propagation between theories

### DISCUSSION THREAD

**Discussion by SpencerL-Y** [cvc5/cvc5#11463] [Q&A]
Hi, recently I am implementing my own theory and the solving process reduce the sat problem to different theory like LIA and SET. So I need to construct the formulas and  propagate these newly created formula to different theory.

I tried the propagation via `d_out->propagate` interface, but since the constraint is newly  created and no literals are mapped to it, this interface cannot be used.   

I also found following comments in `theoryEngine`:
```
/**
   * Literals that are propagated by the theory. Note that these are TNodes.
   * The theory can only propagate nodes that have an assigned literal in the
   * SAT solver and are hence referenced in the SAT solver.
   */
  context::CDList<TNode> d_propagatedLiterals;
```

I wonder if there are existing interfaces that suit my need?  I appreciate that if any help can be provided. Thx

**Comment by ajreynol** [comment]
We unfortunately don't have much public documentation about how to develop theory solvers. We will keep you posted if this changes in the future.

The short answer here: each theory has an "Equality engine" which is the main mechanism for theory propagation for nearly every theory (apart from arithmetic or bv).  To propagate a literal or an equality, it suffices to call EqualityEngine::assertEquality (https://github.com/cvc5/cvc5/blob/main/src/theory/uf/equality_engine.h#L198) or EqualityEngine::assertPredicate on *preregistered* terms. A preregistered term or predicate is one that exists in the SAT solver and thus is relevent for theory propagation. These are notified to Theory::preRegisterTerm (https://github.com/cvc5/cvc5/blob/main/src/theory/theory.h#L409).

All theories are expected to initialize an equality engine, which is managed by methods like https://github.com/cvc5/cvc5/blob/main/src/theory/theory.h#L265.

Again, unfortunately there is no public documentation on how to do this, but this gives a very rough idea.
